[SPARK-26753][CORE] Fixed custom log levels for spark-shell by using Filter instead of Threshold

ankuriitg opened the pull request:
## What changes were proposed in this pull request?

This fix replaces the Threshold with a Filter for ConsoleAppender which checks
to ensure that either the logLevel is greater than thresholdLevel (shell log
level) or the log originated from a custom defined logger. In these cases, it
lets a log event go through, otherwise it doesn't.

## How was this patch tested?

1. Ensured that custom log level works when set by default (via log4j.properties)
2. Ensured that logs are not printed twice when log level is changed by setLogLevel
3. Ensured that custom logs are printed when log level is changed back by setLogLevel

ankuriitg committed:
[SPARK-26753][CORE] Fix for ensuring custom log levels for packages/classes work for spark-shell

This fix replaces the Threshold with a Filter for ConsoleAppender which checks
to ensure that either the logLevel is greater than thresholdLevel (shell log
level) or the log originated from a custom defined logger. In these cases, it
lets a log event go through, otherwise it doesn't.

Testing Done:
1. Ensured that custom log level works when set by default (via log4j.properties)
2. Ensured that logs are not printed twice when log level is changed by setLogLevel
3. Ensured that custom logs are printed when log level is changed back by setLogLevel

ankuriitg commented:
cc @vanzin

AmplabJenkins commented:
Can one of the admins verify this patch?

AmplabJenkins commented:
Can one of the admins verify this patch?

SparkQA commented:
**[Test build #101767 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/101767/testReport)** for PR 23675 at commit [`2a191cf`](https://github.com/apache/spark/commit/2a191cf84cdc44a0e29c672f6d8d7241843d007d).

SparkQA commented:
**[Test build #101767 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/101767/testReport)** for PR 23675 at commit [`2a191cf`](https://github.com/apache/spark/commit/2a191cf84cdc44a0e29c672f6d8d7241843d007d).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/101767/
Test FAILed.

vanzin reviewed (commented):
Please explain the fix, not the problem, in the PR.

ankuriitg committed:
Review comments: Part 1

1. Moved SparkShellLoggingFilter from Utils.scala to Logging.scala
2. Moved thresholdLevel from Filter to Logging.scala
3. Added unit test cases

vanzin commented on core/src/main/scala/org/apache/spark/internal/Logging.scala:
```diff
@@ -213,7 +205,21 @@ private[spark] object Logging {
+              case ssf: SparkShellLoggingFilter =>
+                ca.clearFilters()
+              case f: org.apache.log4j.spi.Filter =>
+                var previous = f
```
I still think the match is not correct. e.g. if the first filter is the one you added, you remove all filters, regardless of whether it's the first one or not.

The code in this `case` correctly handles all possible combinations.

ankuriitg replied:
I just realized that resetting `sparkShellThresholdLevel` is all that I need here (based on your other comment for the test case).

vanzin commented on core/src/main/scala/org/apache/spark/internal/Logging.scala:
```diff
@@ -229,3 +237,33 @@ private[spark] object Logging {
+    }
+    val rootLevel = LogManager.getRootLogger().getLevel()
+    if (loggingEvent.getLevel().isGreaterOrEqual(thresholdLevel) ||
+      !loggingEvent.getLevel().eq(rootLevel)) {
```
Doesn't just `!loggingEvent.getLevel().eq(rootLevel)` cover all cases?

If the log message level is different than the root level, then the only way it can happen is if the logger was customized.

If it's equal, then you need to check whether the logger's level was customized, do decide whether to filter based on the shell threshold.

(BTW the method comment doesn't fully explain the logic here.)

ankuriitg replied:
So, the `loggingEvent.getLevel().isGreaterOrEqual(thresholdLevel)` is just an optimization. If the log message level is same as root level but it is greater than threshold level, then we still allow the event to pass through, irrespective of whether it was customized or not.

vanzin replied:
That's my question. Doesn't `!loggingEvent.getLevel().eq(rootLevel)` already cover that?

The other check is only an optimization if `thresholdLevel` is the same as `rootLevel`, and in that case a better optimization would be to not set `thresholdLevel` (so the null check triggers instead).

Using the good old truth table:

```
Level       Thresh      Root      Action
INFO       WARN       INFO     Check for override
WARN     WARN       INFO     Allow
DEBUG   WARN       INFO     Allow
```

So the resulting action doesn't seem dependent on the threshold level at all, assuming it's always different than the root logger's level.

ankuriitg replied:
Got it. I just checked and `thresholdLevel` is only set when it is different from `rootLevel`. So, this check is not needed and I will remove it.

vanzin commented on core/src/test/scala/org/apache/spark/internal/LoggingSuite.scala:
```diff
@@ -0,0 +1,57 @@
+      Utils.setLogLevel(Level.INFO)
+      assert(ssf.decide(logEvent2) != Filter.DENY)
+    } finally {
+      rootLogger.setLevel(originalLevel)
```
reset `sparkShellThresholdLevel` too?

ankuriitg committed:
Review comments: Part 2

1. Changed uninitialize() method of Logging.scala
2. Other minor modifications

SparkQA commented:
**[Test build #101843 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/101843/testReport)** for PR 23675 at commit [`893ff75`](https://github.com/apache/spark/commit/893ff759cff0df3aa1d2773de21e3f4adcfed0c8).

SparkQA commented:
**[Test build #101836 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/101836/testReport)** for PR 23675 at commit [`7d8fb20`](https://github.com/apache/spark/commit/7d8fb20c8f51713904666c4bddf6e00fba10fed2).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/101836/
Test PASSed.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/101843/
Test PASSed.

vanzin commented on core/src/main/scala/org/apache/spark/internal/Logging.scala:
```diff
@@ -229,3 +218,31 @@ private[spark] object Logging {
   }
 }
+
+private class SparkShellLoggingFilter() extends Filter {
```
remove `()`

vanzin reviewed (commented):
Small nit. I'll fix during merge.

Merging to master.
